File: scripts/strategy_pnl.py
#!/usr/bin/env python3
"""按策略聚合 P&L — 从 reports/trades.jsonl 计算已实现盈亏、胜率、盈亏比"""

import json
import sys
from collections import defaultdict
from pathlib import Path

TRADES_FILE = Path(__file__).parent.parent / "reports" / "trades.jsonl"


def load_trades():
    if not TRADES_FILE.exists():
        return []
    trades = []
    for line in TRADES_FILE.read_text().strip().split("\n"):
        if not line:
            continue
        try:
            trades.append(json.loads(line))
        except json.JSONDecodeError:
            pass
    return trades


def compute_pnl(trades):
    """按策略+symbol分组，配对 buy/sell 计算已实现 P&L。

    简化假设：同策略同 symbol 的交易按时间顺序配对，
    buy 后 sell 为一笔完整交易（或反过来做空）。
    """
    # 按 strategy 分组
    by_strategy = defaultdict(list)
    for t in trades:
        by_strategy[t.get("strategy", "unknown")].append(t)

    results = {}
    for strategy, strades in sorted(by_strategy.items()):
        # 按 symbol 分组内配对
        by_symbol = defaultdict(list)
        for t in strades:
            by_symbol[t.get("symbol", "?")].append(t)

        total_pnl = 0.0
        wins = 0
        losses = 0
        total_win_pnl = 0.0
        total_loss_pnl = 0.0
        trade_count = len(strades)

        for symbol, sym_trades in by_symbol.items():
            # 逐笔配对：记录持仓，遇到反向单算一笔 P&L
            position = None  # {'side': 'buy'|'sell', 'price': float, 'qty': float}
            for t in sym_trades:
                side = t.get("side", "").lower()
                price = float(t.get("price", 0) or 0)
                qty = float(t.get("qty", 0) or 0)

                if price == 0:
                    continue

                if position is None:
                    position = {"side": side, "price": price, "qty": qty}
                    continue

                # 同向加仓：更新均价
                if position["side"] == side:
                    total_qty = position["qty"] + qty
                    if total_qty > 0:
                        position["price"] = (
                            position["price"] * position["qty"] + price * qty
                        ) / total_qty
                        position["qty"] = total_qty
                    continue

                # 反向 → 平仓，计算 P&L
                close_qty = min(position["qty"], qty)
                if position["side"] == "buy":
                    pnl = (price - position["price"]) * close_qty
                else:
                    pnl = (position["price"] - price) * close_qty

                total_pnl += pnl
                if pnl >= 0:
                    wins += 1
                    total_win_pnl += pnl
                else:
                    losses += 1
                    total_loss_pnl += abs(pnl)

                # 剩余持仓
                remaining = position["qty"] - close_qty
                if remaining > 0:
                    position["qty"] = remaining
                elif qty > close_qty:
                    # 反向开仓
                    position = {"side": side, "price": price, "qty": qty - close_qty}
                else:
                    position = None

        round_trips = wins + losses
        win_rate = wins / round_trips if round_trips > 0 else 0
        avg_win = total_win_pnl / wins if wins > 0 else 0
        avg_loss = total_loss_pnl / losses if losses > 0 else 0
        profit_factor = total_win_pnl / total_loss_pnl if total_loss_pnl > 0 else float("inf")

        results[strategy] = {
            "total_pnl": total_pnl,
            "trades": trade_count,
            "round_trips": round_trips,
            "wins": wins,
            "losses": losses,
            "win_rate": win_rate,
            "avg_win": avg_win,
            "avg_loss": avg_loss,
            "profit_factor": profit_factor,
        }

    return results


def main():
    trades = load_trades()
    if not trades:
        print("\n  (无交易记录，reports/trades.jsonl 为空或不存在)")
        return

    results = compute_pnl(trades)

    print(f"\n  {'='*60}")
    print(f"  策略 P&L 聚合  ({len(trades)} 笔交易)")
    print(f"  {'='*60}")

    for strategy, r in sorted(results.items()):
        sign = "+" if r["total_pnl"] >= 0 else ""
        print(f"\n  [{strategy}]")
        print(f"    P&L:       {sign}${r['total_pnl']:.4f}")
        print(f"    交易:      {r['trades']} 笔 ({r['round_trips']} 轮完整)")
        print(f"    胜率:      {r['win_rate']:.0%} ({r['wins']}W/{r['losses']}L)")
        if r["profit_factor"] == float("inf"):
            print(f"    盈亏比:    inf (无亏损)")
        else:
            print(f"    盈亏比:    {r['profit_factor']:.2f}")
        print(f"    平均盈利:  +${r['avg_win']:.4f}")
        print(f"    平均亏损:  -${r['avg_loss']:.4f}")

    # 汇总
    total = sum(r["total_pnl"] for r in results.values())
    total_trades = sum(r["trades"] for r in results.values())
    sign = "+" if total >= 0 else ""
    print(f"\n  {'─'*60}")
    print(f"  合计: {sign}${total:.4f}  ({total_trades} 笔)")
    print()


if __name__ == "__main__":
    main()

File: cli/src/cmd/strategy_pnl.rs
use clawchat_shared::paths::records_dir;
use std::collections::HashMap;

fn parse_num(v: Option<&serde_json::Value>) -> f64 {
    v.and_then(|v| v.as_f64().or_else(|| v.as_str().and_then(|s| s.parse().ok())))
        .unwrap_or(0.0)
}

fn load_trades() -> Vec<serde_json::Value> {
    let path = records_dir().join("trades.jsonl");
    if !path.exists() {
        return Vec::new();
    }
    let content = match std::fs::read_to_string(&path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };
    content
        .lines()
        .filter(|l| !l.is_empty())
        .filter_map(|l| serde_json::from_str(l).ok())
        .collect()
}

struct PnlResult {
    total_pnl: f64,
    trades: usize,
    round_trips: u32,
    wins: u32,
    losses: u32,
    win_rate: f64,
    avg_win: f64,
    avg_loss: f64,
    profit_factor: f64,
}

fn compute_pnl(trades: &[serde_json::Value]) -> HashMap<String, PnlResult> {
    let mut by_strategy: HashMap<String, Vec<&serde_json::Value>> = HashMap::new();
    for t in trades {
        let strat = t
            .get("strategy")
            .and_then(|v| v.as_str())
            .unwrap_or("unknown")
            .to_string();
        by_strategy.entry(strat).or_default().push(t);
    }

    let mut results = HashMap::new();
    for (strategy, strades) in &by_strategy {
        let mut by_symbol: HashMap<String, Vec<&&serde_json::Value>> = HashMap::new();
        for t in strades {
            let sym = t
                .get("symbol")
                .and_then(|v| v.as_str())
                .unwrap_or("?")
                .to_string();
            by_symbol.entry(sym).or_default().push(t);
        }

        let mut total_pnl = 0.0f64;
        let mut wins = 0u32;
        let mut losses = 0u32;
        let mut total_win_pnl = 0.0f64;
        let mut total_loss_pnl = 0.0f64;

        for (_symbol, sym_trades) in &by_symbol {
            let mut pos: Option<(String, f64, f64)> = None; // (side, price, qty)
            for t in sym_trades {
                let side = t
                    .get("side")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_lowercase();
                let price = parse_num(t.get("price"));
                let qty = parse_num(t.get("qty"));
                if price == 0.0 {
                    continue;
                }
                match pos {
                    None => {
                        pos = Some((side, price, qty));
                    }
                    Some((ref ps, pp, pq)) if *ps == side => {
                        let tq = pq + qty;
                        if tq > 0.0 {
                            let np = (pp * pq + price * qty) / tq;
                            pos = Some((side, np, tq));
                        }
                    }
                    Some((ref ps, pp, pq)) => {
                        let close_qty = pq.min(qty);
                        let pnl = if *ps == "buy" {
                            (price - pp) * close_qty
                        } else {
                            (pp - price) * close_qty
                        };
                        total_pnl += pnl;
                        if pnl >= 0.0 {
                            wins += 1;
                            total_win_pnl += pnl;
                        } else {
                            losses += 1;
                            total_loss_pnl += pnl.abs();
                        }
                        let remaining = pq - close_qty;
                        if remaining > 0.0 {
                            pos = Some((ps.clone(), pp, remaining));
                        } else if qty > close_qty {
                            pos = Some((side, price, qty - close_qty));
                        } else {
                            pos = None;
                        }
                    }
                }
            }
        }

        let round_trips = wins + losses;
        let win_rate = if round_trips > 0 {
            wins as f64 / round_trips as f64
        } else {
            0.0
        };
        let avg_win = if wins > 0 {
            total_win_pnl / wins as f64
        } else {
            0.0
        };
        let avg_loss = if losses > 0 {
            total_loss_pnl / losses as f64
        } else {
            0.0
        };
        let profit_factor = if total_loss_pnl > 0.0 {
            total_win_pnl / total_loss_pnl
        } else {
            f64::INFINITY
        };

        results.insert(
            strategy.clone(),
            PnlResult {
                total_pnl,
                trades: strades.len(),
                round_trips,
                wins,
                losses,
                win_rate,
                avg_win,
                avg_loss,
                profit_factor,
            },
        );
    }
    results
}

/// 按策略 P&L — 从交易日志聚合每个策略收益
pub fn strategy_pnl(_days: u32) -> Result<(), Box<dyn std::error::Error>> {
    let trades = load_trades();
    if trades.is_empty() {
        println!("\n  (无交易记录，records/trades.jsonl 为空或不存在)");
        return Ok(());
    }

    let results = compute_pnl(&trades);

    println!("\n  {}", "=".repeat(60));
    println!("  策略 P&L 聚合  ({} 笔交易)", trades.len());
    println!("  {}", "=".repeat(60));

    let mut strats: Vec<_> = results.keys().cloned().collect();
    strats.sort();

    for strategy in &strats {
        let r = &results[strategy];
        let sign = if r.total_pnl >= 0.0 { "+" } else { "" };
        println!("\n  [{strategy}]");
        println!("    P&L:       {sign}${:.4}", r.total_pnl);
        println!(
            "    交易:      {} 笔 ({} 轮完整)",
            r.trades, r.round_trips
        );
        println!(
            "    胜率:      {:.0}% ({}W/{}L)",
            r.win_rate * 100.0,
            r.wins,
            r.losses
        );
        if r.profit_factor.is_infinite() {
            println!("    盈亏比:    inf (无亏损)");
        } else {
            println!("    盈亏比:    {:.2}", r.profit_factor);
        }
        println!("    平均盈利:  +${:.4}", r.avg_win);
        println!("    平均亏损:  -${:.4}", r.avg_loss);
    }

    let total: f64 = results.values().map(|r| r.total_pnl).sum();
    let total_trades: usize = results.values().map(|r| r.trades).sum();
    let sign = if total >= 0.0 { "+" } else { "" };
    println!("\n  {}", "─".repeat(60));
    println!("  合计: {sign}${total:.4}  ({total_trades} 笔)");
    println!();

    Ok(())
}
